plugins/cache_promote/lru_policy.cc

```diff
@@ -72,101 +112,130 @@ LRUPolicy::doPromote(TSHttpTxn txnp)
-      _map.erase(map_it->first);
-      incrementStat(promoted_id, 1);
-      incrementStat(freelist_size_id, 1);
-      decrementStat(lru_size_id, 1);
+      _map.erase(map_key);
+      incrementStat(_promoted_id, 1);
+      incrementStat(_freelist_size_id, 1);
+      decrementStat(_lru_size_id, 1);
       ret = true;
     } else {
       // It's still not promoted, make sure it's moved to the front of the list
-      TSDebug(PLUGIN_NAME, "still not promoted, got %d hits so far", map_it->second->second);
-      _list.splice(_list.begin(), _list, map_it->second);
+      TSDebug(PLUGIN_NAME, "still not promoted, got %d hits so far and %" PRId64 " bytes", val_hits, val_bytes);
+      _list.splice(_list.begin(), _list, map_val);
     }
   } else {
     // New LRU entry for the URL, try to repurpose the list entry as much as possible
-    incrementStat(lru_miss_id, 1);
+    incrementStat(_lru_miss_id, 1);
     if (_list_size >= _buckets) {
       TSDebug(PLUGIN_NAME, "repurposing last LRUHash entry");
       _list.splice(_list.begin(), _list, --_list.end());
-      _map.erase(&(_list.begin()->first));
-      incrementStat(lru_vacated_id, 1);
+      _map.erase(&(std::get<0>(*_list.begin()))); // Get the hash from the first list element
+      incrementStat(_lru_vacated_id, 1);
     } else if (_freelist_size > 0) {
       TSDebug(PLUGIN_NAME, "reusing LRUEntry from freelist");
       _list.splice(_list.begin(), _freelist, _freelist.begin());
       --_freelist_size;
       ++_list_size;
-      incrementStat(lru_size_id, 1);
-      decrementStat(freelist_size_id, 1);
+      incrementStat(_lru_size_id, 1);
+      decrementStat(_freelist_size_id, 1);
     } else {
       TSDebug(PLUGIN_NAME, "creating new LRUEntry");
       _list.push_front(NULL_LRU_ENTRY);
       ++_list_size;
-      incrementStat(lru_size_id, 1);
+      incrementStat(_lru_size_id, 1);
     }
     // Update the "new" LRUEntry and add it to the hash
-    _list.begin()->first          = hash;
-    _list.begin()->second         = 1;
-    _map[&(_list.begin()->first)] = _list.begin();
+    *_list.begin()                       = {hash, 1, 0};
+    _map[&(std::get<0>(*_list.begin()))] = _list.begin();
   }
 
   TSMutexUnlock(_lock);
 
+  // If we didn't promote, and we want to count bytes, save away the calculated hash for later use
+  if (false == ret && countBytes()) {
+    TSUserArgSet(txnp, TXN_ARG_IDX, static_cast<void *>(new LRUHash(hash)));
+  } else {
+    TSUserArgSet(txnp, TXN_ARG_IDX, nullptr);
+  }
+
   return ret;
 }
 
+void
+LRUPolicy::addBytes(TSHttpTxn txnp)
+{
+  LRUHash *hash = static_cast<LRUHash *>(TSUserArgGet(txnp, TXN_ARG_IDX));
+
+  if (hash) {
+    LRUMap::iterator map_it;
+
+    // We have to hold the lock across all list and hash access / updates
+    TSMutexLock(_lock);
+    map_it = _map.find(hash);
+    if (_map.end() != map_it) {
+      TSMBuffer resp;
+      TSMLoc resp_hdr;
+
+      if (TS_SUCCESS == TSHttpTxnServerRespGet(txnp, &resp, &resp_hdr)) {
```

Comment:
You might consider doing this outside of the lock, to reduce contention. In general this method should only be called if it is expected the data will be needed.

Reply:
I decided not to, hoping that TSHttpTxnServerRespGet() etc. are significantly faster then releasing and acquiring the locks again several times would be more costly.